pulser-core/pulser/operators/operator.py

```diff
@@ -0,0 +1,419 @@
+            " should be a tuple of tuples of shape (operator as a "
+            f"{obj_type}, qubit id as an int or str)."
+        )
+    # Check the qubit ids targeted in operations
+    target_qids = [operation[1] for operation in kron_obj]
+    if "global" in target_qids:
+        if len(kron_obj) > 1:
+            raise ValueError(
+                "If a 'global' operation is defined, no other operations "
+                "can be defined."
+            )
+    else:
+        target_qids_list = list(itertools.chain(*target_qids))
+        target_qids_set = set(target_qids_list)
+        if len(target_qids_set) < len(target_qids_list):
+            # Either the qubit id has been defined twice in an operation:
+            for qids in target_qids:
+                if len(set(qids)) < len(qids):
+                    raise ValueError("Duplicate atom ids in argument list.")
+            # Or it was defined in two different operations
+            raise ValueError(
+                "Each qubit can be targeted by only one operation."
+            )
+
+
+@classmethod
+class QuditString:
+    """A 1-qudit state."""
+
+    coefficients: list[Number]
+    states: list[State]
+
+    def __post_init__(self):
+        for state in self.states:
+            if state not in STATES_RANK:
+                raise ValueError(f"State {state} does not exist.")
+
+    def as_tuple(self) -> tuple[tuple[Number, State]]:
+        """Returns tuple of (coefficient, state label)."""
+        return zip(self.coefficients, self.states)
+
+
+@classmethod
+class MultiQuditString:
+    """A multi-qudit state."""
+
+    kron_qudits = tuple[tuple[QuditString, set[QubitId]]]
+
+    def __post_init__(self):
+        _check_kron_object(self.kron_qudits, QuditString)
+
+
+@dataclass
+class QuditOperatorString:
+    """A linear combination of 1-qudit operators."""
+
+    coefficients: list[Number]
+    operators: list[str]
+
+    def __post_init__(self):
+        for operator in self.operators:
+            if operator not in PROJECTORS:
+                raise ValueError(
+                    f"Operators must be among {PROJECTORS}, not {operator}."
+                )
+        if len(set(self.operators)) < len(self.operators):
+            raise ValueError("Operators can only be used once.")
+
+    @property
+    def operations(self) -> tuple[tuple[Number, str]]:
+        """Returns the operations associated with the operator."""
+        return zip(self.coefficients, self.operators)
+
+    def __str__(self) -> str:
+        return str(self.operations)
+
+    def __rmul__(self, scalar: Number) -> QuditOperatorString:
+        return QuditOperatorString(
+            [coeff * scalar for coeff in self.coefficients], self.operators
+        )
+
+    def as_dict(self):
+        return {
+            operator: self.coefficients[i]
+            for (i, operator) in enumerate(self.operators)
+        }
+
+    def __mul__(self, state: QuditString) -> QuditString:
```

Comment:
I have investigated the implementation of multiplication by an operator or a qudit-state, and could generalize this to multi-qudit states, but I would like to know if this is necessary before moving forward @a-quelle @lvignoli

Reply:
It's pretty cool that you're implementing an algebra structure on the abstract representation, but I do wonder if you're not making your life needlessly difficult. I have no immediate use for it, given that I can so far create the state/operator in the backend and add, multiply, apply afterwards. In addition I see the following dificulties:

I imagine that you want the contents of `operators: list[str]` to be essentially arbitrary, since then you can make use of the `dict` in `build_operator` below, but that means that you're going to get stuck with complicated things in your abstract representation for which implementing the algebra structure is hard, and also very slow compared to creating the state in the backend, and doing the operations when things are numbers, and not strings.

Reply:
So I have restricted the operators that can be defined in this QuditOperatorString to projectors, namely {sigma_ab with (a, b) in [u, d, r, g, h]²}, such that I can make this algebra easily. However, I do agree it might be quite slow, plus we don't see yet a clear use of this and it needs quite some work to develop, so I guess it's best to store this as a possible development (I would only need the addition and dagger  operations if I were to develop the Hamiltonian(TimeOperator))